docs: update contributor documentation with new policies (#1329)
dequelabs/cauldron: CONTRIBUTING.md

# Contributing

1. [Getting Started](#getting-started)
   - [Discussing Proprietary Features](#discussing-proprietary-features)
   - [Setup](#setup)
1. [Developing Components](#developing-components)
   - [Tools Used](#tools-used)
   - [Structure](#structure)
   - [Quality Checklist](#quality-checklist)
   - [Server-side Rendering Compatibility](#server-side-rendering-compatibility)
   - [Linting](#linting)
   - [Icons](#icons)
   - [Commit Messages](#commit-messages)
   - [Pull Requests](#pull-requests)
   - [Testing Strategies](#testing-strategies)
   - [Unit Tests](#unit-tests)
   - [Accessibility Testing](#accessibility-testing)
1. [Documentation](#documentation)
1. [Breaking Changes](#breaking-changes)
   - [Components](#components)
   - [Styles](#styles)
   - [Deprecating](#deprectating)
   - [Removal](#removal)

## Getting Started

> Cauldron is Deque's design system / pattern library for building accessible ui as part of our products as Deque. Cauldron consists of several parts, [react](./packages/react), [styles](./packages/styles), and [documentation](./docs).

Reviewer: ```suggestion
Cauldron is Deque's design system/pattern library for building accessible UI as part of our products as Deque. Cauldron consists of several parts, [react](./packages/react), [styles](./packages/styles), and [documentation](./docs).
```

Author: "design system" and "pattern library" are open compounds, so I believe the proper writing style is to include a space between the slash.

Reviewer: Oh yeah, you're right. Grammarly wanted to make it this way. 😄

> As Cauldron is first and foremost a design system for building products at Deque, this means we won't necessarily have 1:1 parity with other design systems. Patterns that don't exist in Deque product ui won't be added as a component within the design system.

Reviewer: ```suggestion
As Cauldron is first and foremost a design system for building products at Deque, this means we won't necessarily have 1:1 parity with other design systems. Patterns that don't exist in Deque product UI won't be added as a component within the design system.
```

### Discussing proprietary features

> This is a public repo, so care needs to be taken to not include details or screenshots for unannounced or unreleased features in Deque products. Any issues or pull requests need to describe work without including specific business logic or feature details directly related to Deque products. If context is necessary to implement a specific Cauldron feature, please open an issue in the private [cauldron-team repo](https://github.com/dequelabs/cauldron-team/issues) and link to it via the issue or pull request.

Reviewer: ```suggestion
This is a public repo, so care needs to be taken to not include details or screenshots for unannounced or unreleased features in Deque products. Any issues or pull requests need to describe work without including specific business logic or feature details directly related to Deque products. If the context is necessary to implement a specific Cauldron feature, please open an issue in the private [cauldron-team repo](https://github.com/dequelabs/cauldron-team/issues) and link to it via the issue or pull request.
```

Author: "If _the_ context is necessary..." sounds really strange to me, as that's makes it seems like there's only one context. I'm thinking of context as a more broad concept here since there could be multiple things that might need to be noted. Does that make sense?

Reviewer: Makes sense to me!

### Setup

Local development setup is documented in [this project's readme](./README.md#development)

## Developing Components

General [component guidelines](https://cauldron.dequelabs.com/component-guidelines) and [style guidelines](https://cauldron.dequelabs.com/style-guidelines) live within our documentation.

- Use well-structured semantic markup over `aria-*` attributes
- Don't specify units for zero values, e.g. `padding: 0px;`, instead you can omit the unit, e.g. `padding: 0;`

> - Avoid using css prefixes (e.g. `-webkit` or `-moz`) unless absolutely necessary. [autoprefixer](https://github.com/postcss/autoprefixer) should automatically handle prefixes.

Reviewer: ```suggestion
- Avoid using CSS prefixes (e.g. `-webkit` or `-moz`) unless absolutely necessary. [autoprefixer](https://github.com/postcss/autoprefixer) should automatically handle prefixes.
```

### Tools Used

- [React](https://react.dev/)
- [Typescript](https://www.typescriptlang.org/)
- [Webpack](https://webpack.js.org/)
- [Jest](https://jestjs.io/)
- [axe-core](https://github.com/dequelabs/axe-core)
- [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/)
- [enzymejs](https://enzymejs.github.io/enzyme)
- [MDXJS](https://mdxjs.com/)
- [ESLint](https://eslint.org/)
- [Prettier](https://prettier.io/)
- [Puppeteer](https://pptr.dev/)

### Structure

```

> ├─ packages/react/src/components
> │  └─ Combobox/
> │     ├─ index.ts           # exports the component and sub components
> │     ├─ Combobox.tsx       # main component file
> │     ├─ ComboboxItem.tsx   # sub component file
> │     └─ Combobox.test.tsx  # unit/behavior tests
> └─ packages/react/styles
>    ├─ index.css             # global styles export
>    └─ combobox.css          # component styles

Reviewer: ```suggestion
├─ packages/react/src/components
│  └─ Combobox/
│     ├─ index.ts           # exports the component and subcomponents
│     ├─ Combobox.tsx       # main component file
│     ├─ ComboboxItem.tsx   # subcomponent file
│     └─ Combobox.test.tsx  # unit/behavior tests
└─ packages/react/styles
   ├─ index.css             # global styles export
   └─ combobox.css          # component styles
```

Cauldron does not have a dedicated quality assurance (QA) individual. Having a full-time dedicated QA individual is a luxury most design systems do not have, ours included. In order to deliver components with high quality it is expected that everyone participating as a part of a component's development lifecycle wear different hats to help ensure we meet that high level of quality. Both an individual developing a component and a reviewer reviewing a component are accountable for considering the following checklist:

- Does the component follow general [component guidelines](https://cauldron.dequelabs.com/component-guidelines)?
- Does the component follow general [style guidelines](https://cauldron.dequelabs.com/style-guidelines)?
- Does the component cover all of the necessary variations within scope? (i.e. default, primary, secondary)
- Does the component cover all of the necessary interactive states? (i.e. active, focus, expanded, disabled)
- Does the component account for different layouts within different viewports? (i.e. responsiveness, media queries, overflow)
- Does the component api match the current design patterns of components? (i.e. props, events, callbacks)

> - Does the component only implement features that fall within our browser support, or implement these features via progressive enhancement? (TODO: determine browser support)

Reviewer: ```suggestion
- Does the component only implement features that fall within our browser support or implement these features via progressive enhancement? (TODO: determine browser support)
```

|                      | A required prop is added                                              | `major` (breaking) |
|                      | A prop's type is expanded                                             | `minor`            |
|                      | A prop's type is narrowed                                             | `major` (breaking) |
|                      | A prop is deprecated                                                  | `minor`            |
|                      | A prop is removed                                                     | `major` (breaking) |
| Component Styles/CSS | Adding a new css selector                                             | `minor`            |
|                      | Removing a css selector                                               | `major` (breaking) |
|                      | Adding a css variable                                                 | `minor`            |
|                      | Removing a css variable                                               | `major` (breaking) |
|                      | Modifying css properties                                              | `patch` or `minor` |

### Pull Requests

> We aim to review pull requests within a day or two. If your changes are time sensitive, feel free to request an expedited review on GitHub or reach out in the private `#cauldron` slack channel.

Reviewer: ```suggestion
We aim to review pull requests within a day or two. If your changes are time-sensitive, feel free to request an expedited review on GitHub or reach out in the private `#cauldron` slack channel.
```

Once approved by a member of the Cauldron team, your pull request can be merged at any time but _must_ be squash merged.

#### Previewing Changes

> Cauldron documentation is deployed automatically via [amplify web previews for pull requests](https://docs.aws.amazon.com/amplify/latest/userguide/pr-previews.html) with each commit to a PR. To view the preview of your changes, navigate to your PR and find the comment from the `aws-amplify` bot, which will include a link to the preview site for that PR. The preview site will only persist for as long as the PR remains opened, and will be deleted when closed.

Reviewer: ```suggestion
Cauldron documentation is deployed automatically via [amplify web previews for pull requests](https://docs.aws.amazon.com/amplify/latest/userguide/pr-previews.html) with each commit to a PR. To view the preview of your changes, navigate to your PR and find the comment from the `aws-amplify` bot, which will include a link to the preview site for that PR. The preview site will only persist for as long as the PR remains opened and will be deleted when closed.
```

### Testing Strategies

> Testing methodology should account for both testing the interface of a component as input/output testing and with how an end-user would interact with the component. Ideally, implementation details are avoided unless necessary. Examples below of things to consider when writing tests for a component:

Reviewer: ```suggestion
Testing methodology should account for both testing the interface of a component as input/output testing and how an end-user would interact with the component. Ideally, implementation details are avoided unless necessary. Examples below of things to consider when writing tests for a component:
```

- Including all variants in tests (i.e. default, primary, secondary)
- Including different interactive states (i.e. focus, expanded, disabled)

> - Including testing different behaviors for functionality related to different layouts (i.e. i.e. responsiveness, media queries, overflow)

Reviewer: ```suggestion
- Including testing different behaviors for functionality related to different layouts (i.e. responsiveness, media queries, overflow)
```

> - Component api and events (i.e. callbacks, click, keyboard events)

Reviewer: ```suggestion
- Component API and events (i.e. callbacks, clicks, keyboard events)
```

Author: This was supposed to be referencing click / keyboard events, so perhaps that's what it should be changed to?

Reviewer: So "i.e. callbacks, click / keyboard events"?

Author: Correct, made the change!

- Running axe against various variants and states

### Unit Tests

> Cauldron uses [Jest](https://jestjs.io/) as its test runner to run unit tests along with [enzymejs](https://enzymejs.github.io/enzyme) (deprecated) and [react testing library](https://testing-library.com/docs/react-testing-library/intro/). We are currently in the process of migrating tests away from enzyme to react testing library so no new tests should be using enzyme.

Reviewer: ```suggestion
Cauldron uses [Jest](https://jestjs.io/) as its test runner to run unit tests along with [enzymejs](https://enzymejs.github.io/enzyme) (deprecated) and [react testing library](https://testing-library.com/docs/react-testing-library/intro/). We are currently in the process of migrating tests away from enzyme to React Testing Library so no new tests should be using Enzyme.
```
